### PR TITLE
Make connection thread safe

### PR DESCRIPTION
**Pull request by swember**
`@connection` class variable in `ChartMogul::APIResource` caching connection credentials which lead to an unexpected result if credentials are changed in runtime. This PR makes `@connection` threadsafe and clears it if credentials are changed.

### CODE REVIEW

File: lib/chartmogul/api_resource.rb [hunk 1]
```diff
@@ -31,14 +32,8 @@ def self.set_resource_root_key(root_key)
...
     def self.connection
-      @connection ||= Faraday.new(url: ChartMogul::API_BASE) do |faraday|
-        faraday.use Faraday::Request::BasicAuthentication, ChartMogul.account_token, ChartMogul.secret_key
-        faraday.use Faraday::Response::RaiseError
-        faraday.request :retry, max: ChartMogul.max_retries, retry_statuses: RETRY_STATUSES,
-                                max_interval: MAX_INTERVAL, backoff_factor: BACKOFF_FACTOR,
-                                interval_randomness: INTERVAL_RANDOMNESS, interval: INTERVAL, exceptions: RETRY_EXCEPTIONS
-        faraday.use Faraday::Adapter::NetHttp
-      end
+      Thread.current["#{THREAD_PREFIX}.connection"] = build_connection if creds_changed?
```
Comment (pkopac) [inline lib/chartmogul/api_resource.rb:35]: Is this intentional: `chartmogul_ruby.connection.connection` ?
What about instead - when you change configuration, you `nil` this thread variable? Would it make simpler?
Ie. invalidate the cached connection when setting credentials.

File: spec/chartmogul/api_resource_spec.rb [hunk 2]
```diff
@@ -3,19 +3,42 @@
...
       set_invalid_credentials
       expect { ChartMogul::Ping.ping }.to raise_error(ChartMogul::UnauthorizedError)
 
       set_valid_credentials
-      expect { ChartMogul::Ping.ping }.not_to raise_error(ChartMogul::UnauthorizedError)
+      expect { ChartMogul::Ping.ping }.not_to raise_error
+    end
+
+    it 'works in a threaded environment', vcr: { record: :all, exclusive: true } do
+      t1 = Thread.new do
```
Comment (pkopac) [inline spec/chartmogul/api_resource_spec.rb:16]: This is starting two new threads and joining into the main one later. The order of their execution on a 1-core CPU is random, to make it a bit more predictable, it's better to set something in the main thread and then run another thread (join).

File: spec/chartmogul/api_resource_spec.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    end
+
+    it 'works in a threaded environment' do
+      set_invalid_credentials
+      expect { ChartMogul::Ping.ping }.to raise_error(ChartMogul::UnauthorizedError)
+
+      Thread.new do
+        set_valid_credentials
+        expect { ChartMogul::Ping.ping }.not_to raise_error
+      end.join
```
Comment (pkopac) [inline spec/chartmogul/api_resource_spec.rb:23]: and now add another `expect { ChartMogul::Ping.ping }.to raise_error(ChartMogul::UnauthorizedError)` in the main thread without setting again the credentials, which will confirm that the connection was indeed invalidated.

**Comment by pkopac** [comment]
@swember I almost forgot. You need to bump up the version to 1.6.1, since we've already released the faulty one as 1.6.0.

**Comment by swember** [comment]
> @swember I almost forgot. You need to bump up the version to 1.6.1, since we've already released the faulty one as 1.6.0.

@pkopac, done

**Pull request merged**